src/librustdoc/json/mod.rs

```diff
@@ -1,47 +1,238 @@
+        item.inner.inner_items().for_each(|i| self.item(i.clone(), cache).unwrap());
+
+        let id = item.def_id;
+        let mut new_item: types::Item = item.into();
+        if let types::ItemEnum::TraitItem(ref mut t) = new_item.inner {
+            t.implementors = self.get_trait_implementors(id, cache)
+        } else if let types::ItemEnum::StructItem(ref mut s) = new_item.inner {
+            s.impls = self.get_impls(id, cache)
+        } else if let types::ItemEnum::EnumItem(ref mut e) = new_item.inner {
+            e.impls = self.get_impls(id, cache)
+        }
+
+        self.index.borrow_mut().insert(id.into(), new_item);
+        Ok(())
     }
 
     fn mod_item_in(
         &mut self,
-        _item: &clean::Item,
-        _item_name: &str,
-        _cache: &Cache,
+        item: &clean::Item,
+        _module_name: &str,
+        cache: &Cache,
     ) -> Result<(), Error> {
-        unimplemented!()
+        use clean::types::ItemEnum::*;
+        if let ModuleItem(m) = &item.inner {
+            for item in &m.items {
+                match &item.inner {
+                    // These don't have names so they don't get added to the output by default
+                    ImportItem(_) => self.item(item.clone(), cache).unwrap(),
+                    ExternCrateItem(_, _) => self.item(item.clone(), cache).unwrap(),
+                    ImplItem(i) => {
+                        i.items.iter().for_each(|i| self.item(i.clone(), cache).unwrap())
+                    }
+                    _ => {}
+                }
+            }
+        }
+        self.item(item.clone(), cache).unwrap();
+        Ok(())
     }
 
     fn mod_item_out(&mut self, _item_name: &str) -> Result<(), Error> {
-        unimplemented!()
+        Ok(())
     }
 
-    fn after_krate(&mut self, _krate: &clean::Crate, _cache: &Cache) -> Result<(), Error> {
-        unimplemented!()
+    fn after_krate(&mut self, krate: &clean::Crate, cache: &Cache) -> Result<(), Error> {
+        debug!("Done with crate");
+        let mut index = (*self.index).clone().into_inner();
+        index.extend(self.get_trait_items(cache));
+        let output = types::Crate {
+            root: types::Id(String::from("0:0")),
+            version: krate.version.clone(),
+            includes_private: cache.document_private,
+            index,
+            paths: cache
+                .paths
+                .clone()
+                .into_iter()
+                .chain(cache.external_paths.clone().into_iter())
+                .map(|(k, (path, kind))| {
+                    (
+                        k.into(),
+                        types::ItemSummary { crate_id: k.krate.as_u32(), path, kind: kind.into() },
+                    )
+                })
+                .collect(),
+            external_crates: cache
+                .extern_locations
+                .iter()
+                .map(|(k, v)| {
+                    (
+                        k.as_u32(),
```

Comment:
Having numeric keys seems annoying. Would it make more sense to instead have a list that you look up by index?

Concretely, currently the output looks like this:

```
{
    "1": {
      "name": "std",
      "html_root_url": "https://doc.rust-lang.org/nightly/"
    },
    "2": {
      "name": "core",
      "html_root_url": "https://doc.rust-lang.org/nightly/"
    },
  ...
}
```

and I'd prefer it to look like

```
[
    {
      "name": "std",
      "html_root_url": "https://doc.rust-lang.org/nightly/"
    },
   {
      "name": "core",
      "html_root_url": "https://doc.rust-lang.org/nightly/"
    }
  ...
]
```